File: Econovation_Algorithm_Mentoring/1. 완전 탐색 (Brute-force Search)/Programmers_PrimeNumber.java
import java.util.HashSet;

public class Programmers_PrimeNumber {
    static int[] arr;
    static int count;
    HashSet<Integer> hs;

    public int solution(String numbers) {
        arr = new int[numbers.length()];
        for (int i = 0; i < numbers.length(); i++) {
            arr[i] = numbers.charAt(i) - '0';
        }

        count = 0;
        hs = new HashSet<>();
        for (int i = 1; i <= numbers.length(); i++) {
            int[] temp = new int[i];
            boolean visit[] = new boolean[arr.length];
            makeCombination(i, temp, 0, visit);
        }
        return count;
    }

    public void makeCombination(int r, int[] temp, int current, boolean[] visit) {
        if (r == current) {
            String num = "";
            for (int i = 0; i < temp.length; i++) {
                num += temp[i];
            }
            if (num.charAt(0) == '0' && num.length() > 1) {
                num = num.substring(1);
            }
            int number = Integer.parseInt(num);

            if (number == 1 || number == 0) {
                return;
            }

            if (hs.contains(number)) {
                return;
            } else {
                hs.add(number);
            }

            if (number == 2) {
                count++;
                return;
            }
            
            for (int i = 2; i <= Math.sqrt(number); i++) {
                if (number % i == 0) {
                    return;
                }
            }
            count++;
        } else {
            for (int i = 0; i < arr.length; i++) {
                if (!visit[i]) {
                    visit[i] = true;
                    temp[current] = arr[i];
                    makeCombination(r, temp, current + 1, visit);
                    visit[i] = false;
                }

            }
        }
    }
}

File: Algorithm_Swift/Algorithm_Swift/Programmers/Programmers_PrimeNumber.swift
//
//  Programmers_PrimeNumber.swift
//  Algorithm_Swift
//
//  Created by asong on 2022/05/23.
//

import Foundation
func PrimeNumber(){
    var answer = 0
    var set:Set<Int> = Set<Int>()
    func solution(_ numbers:String) -> Int {
        let num = numbers.map{String($0)}
        
        for i in 1...numbers.count{
            let temp = Array(repeating: "", count: i)
            let visit = Array(repeating: false, count: num.count)
            makePermutation(i,0,num,visit, temp)
        }
        return answer
    }
    func makePermutation(_ r: Int,_ current: Int, _ arr: [String], _ visit: [Bool], _ temp: [String]){
        var temp = temp
        var visit = visit
        
        if r == current{
            var str = ""
            for i in 0..<temp.count{
                str += temp[i]
            }
            let nn = Int(str)!
            if checkPrimeNumber(nn){
                if !set.contains(nn){
                    answer += 1
                    set.insert(nn)
                }
            }
        }else{
            for i in 0..<arr.count{
                if !visit[i]{
                    visit[i] = true
                    temp[current] = arr[i]
                    makePermutation(r, current+1, arr, visit, temp)
                    visit[i] = false
                }
            }
        }
    }
    func checkPrimeNumber(_ n:Int) -> Bool{
        if n <= 1{
            return false
        }
        if n == 2{
            return true
        }
        if n%2 == 0{
            return false
        }
        var i = 3
        while i <= Int(sqrt(Float(n)))+1{
            if n % i == 0{
                return false
            }
            i += 2
        }
        return true
    }

}

